### PR TITLE
Feature/modernize cmake

### PR DESCRIPTION
**Pull request by ClausKlein**
Use `FILE_SET HEADERS` for simplify installation.

Use `VERIFY_INTERFACE_HEADER_SETS`

Prevent use of `include(CTest)`.

Prepare CPack to easy distribution. 

- [x] depends on https://github.com/bemanproject/inplace_vector/pull/42

NOTE: This is first of a series of PR based on https://github.com/bemanproject/inplace_vector/pull/14

This fix https://github.com/bemanproject/inplace_vector/issues/35

### CODE REVIEW

File: include/beman/inplace_vector/inplace_vector.hpp
Comment (DeveloperPaul123) [inline include/beman/inplace_vector/inplace_vector.hpp:77]: Could you remove this change just to keep the scope of this PR to CMake/infrastructure?

**Review by DeveloperPaul123 — CHANGES_REQUESTED** [review]
Just a minor comment to address.

File: CMakeLists.txt
Comment (ClausKlein) [reply CMakeLists.txt:63]: But this exports your min required `cxx_std` on this machine.

File: include/beman/inplace_vector/inplace_vector.hpp
Comment (ClausKlein) [reply include/beman/inplace_vector/inplace_vector.hpp:77]: Then it does not compile!

File: CMakeLists.txt
Comment (DeveloperPaul123) [reply CMakeLists.txt:63]: `inplace_vector` targets a min standard of C++17 I believe.

Comment (ClausKlein) [reply CMakeLists.txt:63]: **Note from CI workflow**: Needs C++ 20 as C++17 selectively disables ranges and concepts related functionalities!

**Comment by ClausKlein** [comment]
@wusatosi This PR was requested from you. Any notes?

File: CMakeLists.txt
Comment (wusatosi) [reply CMakeLists.txt:63]: CK is right here

Comment (wusatosi) [reply CMakeLists.txt:63]: CK is right here regarding the C++ version.

**Comment by wusatosi** [comment]
> @wusatosi This PR was requested from you. Any notes?

Thanks! You made this a lot easier to review :P

Thank you for taking my feedback. I am not familiar with this part of CMake

I will leave the review up to @DeveloperPaul123 .

**Review by wusatosi — APPROVED** [review]
I confirm this PR fixes the install command and enables all_verify_interface_header_sets.

Nice job @ClausKlein , I will leave the details to @DeveloperPaul123 .

File: CMakeLists.txt
Comment (DeveloperPaul123) [reply CMakeLists.txt:63]: Ok that's fine, but what about setting the flags on the target to begin with? I thought we were not setting these properties on build/header targets so that they don't propagate downstream to our consumers? @wusatosi

Comment (DeveloperPaul123) [reply CMakeLists.txt:63]: I would recommend to move setting of the feature flags to the unit test target instead of the header library (`INTERFACE` target).

Comment (wusatosi) [reply CMakeLists.txt:63]: Ah I see what's going on here.

Yes we discourage use of `target_compile_features`.

See: https://github.com/bemanproject/exemplar/issues/40#issuecomment-2388995030 , https://github.com/bemanproject/optional26/pull/80

Comment (ClausKlein) [reply CMakeLists.txt:63]: I see no problem, and it may be changed later to any decision.

Comment (ClausKlein) [reply CMakeLists.txt:63]: @wusatosi may it possible to finish this **PR** right now, I want to continue with the `cmake presets`

Comment (wusatosi) [reply CMakeLists.txt:63]: sorry I cannot include this practice that we knowingly discourage & reject. 

I don't think you've shown that your solution doesn't address the original concern. target_compile_features may lead to ABI break & ODR violation **at link time** as documented in previous reference.

This is also documented elsewhere, see:
https://cgold.readthedocs.io/en/latest/rejected/target-compile-features.html#target-compile-features
https://gitlab.kitware.com/cmake/cmake/-/issues/19136

Comment (ClausKlein) [reply CMakeLists.txt:63]: I am sure you are misunderstanding the Problem:
```cmake
# Usage:
#   cmake -G Ninja -S . -B build -D CMAKE_CXX_STANDARD=98 --fresh
#   ninja -C build -v
#   build/exe

cmake_minimum_required(VERSION 3.30...3.31)

project(Test LANGUAGES CXX)

add_library(lib1 lib1.cpp)
target_compile_features(lib1 PUBLIC cxx_std_14)

add_library(lib2 lib2.cpp)
target_compile_features(lib2 PUBLIC cxx_std_11)
# ORIG: target_link_libraries(lib2 PRIVATE lib1)
target_link_libraries(lib2 PUBLIC lib1)

add_executable(exe exe.cpp)
target_link_libraries(exe PRIVATE lib2)
```

As long each static library is build within one build step, you can control it.
### But not if the libraries are build independently and installed without a cmake  config package!

And for a header only library, this ODR-violation can't happen!

```bash
bash-5.2$ CXX=g++ cmake -G Ninja -S . -B build -D CMAKE_CXX_STANDARD=98 --fresh
-- The CXX compiler identification is AppleClang 16.0.0.16000026
-- Detecting CXX compiler ABI info
-- Detecting CXX compiler ABI info - done
-- Check for working CXX compiler: /Applications/Xcode.app/Contents/Developer/usr/bin/g++ - skipped
-- Detecting CXX compile features
-- Detecting CXX compile features - done
-- Configuring done (16.0s)
-- Generating done (0.0s)
-- Build files have been written to: /Users/clausklein/Downloads/ODR-violoation/build
bash-5.2$ build/exe
lib1: 201402
lib2: 201402
exe: 201402
bash-5.2$ 
```

Comment (ClausKlein) [reply CMakeLists.txt:63]: I am waiting for a pragmatic decision ...

Comment (wusatosi) [reply CMakeLists.txt:63]: Mentioned... waiting for someone to come in and review....

File: .gitignore
Comment (camio) [inline .gitignore:5]: Why these files?

File: CMakeLists.txt
Comment (wusatosi) [reply CMakeLists.txt:20]: If this is not desired across beman, we should close https://github.com/bemanproject/exemplar/issues/67 .

Comment (ClausKlein) [reply CMakeLists.txt:20]: It is an open issue at CMake to deny `in-source builds.`

if you do `cmake . -L`  in _source_dir_ instead in _binary_dir_, please note the generated files in source tree!

File: CMakeLists.txt
Comment (ClausKlein) [reply CMakeLists.txt:36]: This is no restriction, he user has still the choice to use `cpack -G ZIP ...`

File: CMakeLists.txt
Comment (ClausKlein) [reply CMakeLists.txt:62]: see my example https://github.com/bemanproject/inplace_vector/pull/33#discussion_r1861512422

File: CMakeLists.txt
Comment (ClausKlein) [reply CMakeLists.txt:6]: With this setting, it is possible to do `make test `without `make all` before.

File: CMakeLists.txt
Comment (ClausKlein) [reply CMakeLists.txt:8]: To prevent policy warnings if you use a newer CMake version than `v.3.25`

I would use `3.30...3.31` today.

File: .gitignore
Comment (ClausKlein) [reply .gitignore:5]: Garbage generated be in-source build which `cmake .`

File: CMakeLists.txt
Comment (ClausKlein) [reply CMakeLists.txt:45]: I have to check ...

File: CMakeLists.txt
Comment (wusatosi) [reply CMakeLists.txt:20]: @ClausKlein please stop falsely marking conversations as resolved.

File: CMakeLists.txt
Comment (camio) [reply CMakeLists.txt:6]: I'd like to see how my [suggestion to change the default in CMake](https://discourse.cmake.org/t/cmake-skip-test-all-dependency-has-the-wrong-default/13144) plays out. If they're okay with changing the default, I'd be okay with leaving this line in as a temporary solution.

Comment (wusatosi) [reply CMakeLists.txt:6]: David, this series of changes has been pending for half a month plus (it was originally a big pr that was broken down into smaller ones).

Can we have a review direction here first and remove/ add back in after there's discussion on the cmake forum. 

Due to the fact that this is a piece of an originally bigger PR, I believe there are other pending PRs that are based on changes here CK wishes to contribute.

Let's not block the review queue by waiting for an external CMake forum.

File: .gitignore
Comment (camio) [reply .gitignore:5]: I don't think we should add to our `.gitignore` to cover cases where users do unsupported things.

Every line of code we add has an associated maintenance cost and an overhead for readers. I don't think, in this instance, the cost justifies the benefit.

File: CMakeLists.txt
Comment (camio) [reply CMakeLists.txt:6]: The discussion there has lead me the conclusion that it is not a good practice to use `set(CMAKE_SKIP_TEST_ALL_DEPENDENCY FALSE)`. See [this comment](https://discourse.cmake.org/t/cmake-skip-test-all-dependency-has-the-wrong-default/13144/6?u=camio) for a rationale summary.

Comment (ClausKlein) [reply CMakeLists.txt:6]: That is your interpretation, not a fact!

File: .gitignore
Comment (ClausKlein) [reply .gitignore:5]: see https://github.com/bemanproject/inplace_vector/pull/33#discussion_r1872664908

File: CMakeLists.txt
Comment (ClausKlein) [reply CMakeLists.txt]: Note [Some Professional CMake recommendations](https://discourse.bemanproject.org/t/c-20-modules-in-beman-project-libraies/279/5?u=clausklein)

File: CMakeLists.txt
Comment (camio) [reply CMakeLists.txt:51]: I don't see a benefit of changing the default here, but I see a cost in complexity.

File: CMakeLists.txt
Comment (camio) [reply CMakeLists.txt:60]: I understand that. See what is [done in exemplar](https://github.com/bemanproject/exemplar/blob/589d0f81a505721e67de8a505bd14ee74ffaf8c2/src/beman/exemplar/CMakeLists.txt#L8).

File: .gitignore
Comment (camio) [reply .gitignore:5]: I'm still unconvinced adding these lines is worth the complexity.

File: CMakeLists.txt
Comment (ClausKlein) [reply CMakeLists.txt:6]: not possible!

**Pull request closed without merging**
